winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,198 @@
+	      buffer, strlen (buffer), NULL, NULL);
+}
+
+static int
+inject_remote_thread_into_process(HANDLE process, LPTHREAD_START_ROUTINE address, uintptr_t exit_code)
+{
+  int res = -1;
+
+  if (!address)
+    return res;
+  DWORD thread_id;
+  HANDLE thread = CreateRemoteThread(process, NULL, 1024 * 1024, address,
+                                     (PVOID)exit_code, 0, &thread_id);
+  if (thread)
+  {
+    /*
+      * Wait up to 10 seconds (arbitrary constant) for the thread to finish;
+      * After that grace period, fall back to terminating non-gently.
+      */
+    if (WaitForSingleObject(thread, 10000) == WAIT_OBJECT_0)
+      res = 0;
+    CloseHandle(thread);
+  }
+
+  return res;
+}
+
+static WINAPI BOOL
+ctrl_handler(DWORD ctrl_type)
+{
+  unsigned short count;
+  void *address;
+  HANDLE process;
+  PSYMBOL_INFOW info;
+  DWORD64 displacement;
+
+  count = CaptureStackBackTrace (1l /* skip this function */,
+			         1l /* return only one trace item */,
+				 &address, NULL);
+  if (count != 1)
+    {
+      output (1, "Could not capture backtrace\n");
+      return FALSE;
+    }
+
+  process = GetCurrentProcess ();
+  if (!SymInitialize (process, NULL, TRUE))
+    {
+      output (1, "Could not initialize symbols\n");
+      return FALSE;
+    }
+
+  info = (PSYMBOL_INFOW)
+    malloc (sizeof(*info) + MAX_SYM_NAME * sizeof(wchar_t));
+  if (!info)
+    {
+      output (1, "Could not allocate symbol info structure\n");
+      return FALSE;
+    }
+  info->SizeOfStruct = sizeof(*info);
+  info->MaxNameLen = MAX_SYM_NAME;
+
+  if (!SymFromAddrW (process, (DWORD64)(intptr_t)address, &displacement, info))
+    {
+      output (1, "Could not get symbol info\n");
+      SymCleanup(process);
+      return FALSE;
+    }
+
+  if (do_print_address)
+  {
+    output(0, "%p\n", (void *)(intptr_t)info->Address);
+  }
+  else
+  {
+    HANDLE main_process = OpenProcess(PROCESS_CREATE_THREAD |
+                                          PROCESS_QUERY_INFORMATION |
+                                          PROCESS_VM_OPERATION | PROCESS_VM_WRITE |
+                                          PROCESS_VM_READ,
+                                      FALSE, pid);
+    if (main_process == NULL)
+      return 1;
+    /* Inject the remote thread only when asked too*/
+    inject_remote_thread_into_process(main_process, (LPTHREAD_START_ROUTINE)(intptr_t)info->Address, exit_code);
+  }
+  CloseHandle(GetStdHandle(STD_OUTPUT_HANDLE));
+  SymCleanup(process);
+
+  exit(0);
+}
+
+int main(int argc, char **argv)
+{
+    char *end;
+    if (argc != 2 && argc < 4)
+    {
+        output(1, "Need a function name, exit code and pid\n");
+        output(1, "Or needs a function name.\n");
+        return 1;
+    }
+    if (argc == 4)
+    {
+        exit_code = atoi(argv[2]);
+        pid = strtoul(argv[3], NULL, 0);
+    }
+    if (argc == 2)
+    {
+        do_print_address = TRUE;
+    }
+
+    if (strcmp(argv[1], "CtrlRoutine"))
```

Comment:
This would be a fine place to say that the easy route is to use `GetProcAddress()` (and that this route is unfortunately not viable for `CtrlRoutine` because that symbol is not exported).

Reply:
Done in https://github.com/msys2/msys2-runtime/pull/31/commits/9489d1ad2401b33cf1b9d881329fbb2a76f4945c